Weighted

hgloeckner opened the pull request:
added the weighted functions :) 

(rebased on #165 )e

It's quite a lot... If you like I can try splitting the distance and autocorrelation calculation

tmieslinger reviewed (commented):
I am having a hard time testing this PR. From going through the code I have one suggestion, but other than that, the code seems fine. 
Maybe @d70-t could have a look on this PR?

d70-t commented on pydropsonde/processor.py:
```diff
@@ -2146,6 +2147,122 @@ def add_l3_ds(self, l3_dir: str = None):
+                )
+            elif method == "no_weights":
+                weights = self.get_no_change_weights(variable=var)
+            weights = xr.where(weights < 0, 0, weights)
+            weights = xr.where(weights > 1, 1, weights)
```
```suggestion
            weights = weights.clip(0, 1)
```

hgloeckner commented:
Thank you so much for your suggestions and comments! I think, I incorporated all of them. The main changes to the state before are: 

- correlation is calculated using SciPy.signal.correlate
- autocorrelation functions are split more and the functions to get autocorrelation and distances are outsourced to `helper.stats`
- I removed the weighting and autocorrelation part from the standard pipeline - we decided that it's a nice option for tests, but that for ORCESTRA soundings it is not necessary and hence removed. 
- the simplifications/changes for the weighted fit are also incorporated

d70-t commented on pydropsonde/circles.py:
```diff
@@ -219,33 +219,40 @@ def interpolate_na_sondes(self, method="akima", max_gap=1500, thresh=4):
-        u_cal = np.where(invalid, 0, u)
         a[invalid] = 0
+        sqw = np.where(invalid, 0, np.sqrt(weight))
+        wa = sqw[..., np.newaxis] * a  # np.einsum("...m,...mr->...mr", w, a)
+        u_cal = sqw * np.where(
+            invalid, 0, u
+        )  # np.einsum("...m,...m->...m", w, np.where(invalid, 0, u))
```
I'd suggest to remove the comments and pick either implementation. (The comments don't really explain anything not written in the code, and actually seem to do something different)

d70-t reviewed (commented):
LGTM :+1:
